# Sorting, Counting, Filtering & Querying

## Upload Olympics datasets to this notebook.

### Import ```summer.csv``` and store in a dataframe called ```dfs``` (<a href="https://raw.githubusercontent.com/sandeepmj/datasets/main/summer.csv">link</a>).


### Import ```winter.csv``` and store in a dataframe called ```dfw``` (<a href="https://raw.githubusercontent.com/sandeepmj/datasets/main/winter.csv">link</a>).

In [2]:
#import pandas
import pandas as pd

In [4]:
## import summer data
dfs = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/main/summer.csv")
dfs.head(3)

Year    City     Sport Discipline       Athlete Name Athlete Country  \
0  1896  Athens  Aquatics   Swimming      HAJOS, Alfred             HUN   
1  1896  Athens  Aquatics   Swimming   HERSCHMANN, Otto             AUT   
2  1896  Athens  Aquatics   Swimming  DRIVAS, Dimitrios             GRE   

  Gender                       Event   Medal  
0    Men              100M Freestyle    Gold  
1    Men              100M Freestyle  Silver  
2    Men  100M Freestyle For Sailors  Bronze

In [5]:
## import summer data
dfw = pd.read_csv("https://raw.githubusercontent.com/sandeepmj/datasets/main/winter.csv")
dfw.head(3)

Year      City     Sport Discipline         Athlete Name Athlete Country  \
0  1924  Chamonix  Biathlon   Biathlon          BERTHET, G.             FRA   
1  1924  Chamonix  Biathlon   Biathlon       MANDRILLON, C.             FRA   
2  1924  Chamonix  Biathlon   Biathlon  MANDRILLON, Maurice             FRA   

  Gender            Event   Medal  
0    Men  Military Patrol  Bronze  
1    Men  Military Patrol  Bronze  
2    Men  Military Patrol  Bronze

## Get a sense of the data

In [6]:
## SUMMER
## what exactly do we have: columns, datatypes, etc.
dfs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31165 entries, 0 to 31164
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Year             31165 non-null  int64 
 1   City             31165 non-null  object
 2   Sport            31165 non-null  object
 3   Discipline       31165 non-null  object
 4   Athlete Name     31165 non-null  object
 5   Athlete Country  31161 non-null  object
 6   Gender           31165 non-null  object
 7   Event            31165 non-null  object
 8   Medal            31165 non-null  object
dtypes: int64(1), object(8)
memory usage: 2.1+ MB


In [7]:
dfw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5770 entries, 0 to 5769
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Year             5770 non-null   int64 
 1   City             5770 non-null   object
 2   Sport            5770 non-null   object
 3   Discipline       5770 non-null   object
 4   Athlete Name     5770 non-null   object
 5   Athlete Country  5770 non-null   object
 6   Gender           5770 non-null   object
 7   Event            5770 non-null   object
 8   Medal            5770 non-null   object
dtypes: int64(1), object(8)
memory usage: 405.8+ KB


#### Multiple ```series``` make up a ```dataframe```

The key is to know what you are working with because once again, there are specific operations you can run on a ```series``` that you can't run on a ```dataframe```, and vise versa.

### Method 1 to return series: dot notation

```dataframe.column_name```

In [11]:
## return Year as a Pandas Series for winter games
dfw.Year

0       1924
1       1924
2       1924
3       1924
4       1924
        ... 
5765    2014
5766    2014
5767    2014
5768    2014
5769    2014
Name: Year, Length: 5770, dtype: int64

### Method 2 to return series: square brackets

```dataframe["column_name"]```

In [13]:
## return Year as a Pandas Series for winter games
dfw["Year"]

0       1924
1       1924
2       1924
3       1924
4       1924
        ... 
5765    2014
5766    2014
5767    2014
5768    2014
5769    2014
Name: Year, Length: 5770, dtype: int64

### Method 2 is preferred because you will often work with data with multiword column headers:

This will break:

```df.Multi Word Header```

This will NOT break:

```df["Multi Word Header"]```


In [17]:
## Call the winter games Athlete Name series using dot notation
# will break
dfw.Athlete Name

SyntaxError: invalid syntax (<ipython-input-17-875fa948d3ba>, line 3)

In [18]:
## Call the winter games Athlete Name series using dot notation
dfw["Athlete Name"]

0                BERTHET, G.
1             MANDRILLON, C.
2        MANDRILLON, Maurice
3            VANDELLE, André
4       AUFDENBLATTEN, Adolf
                ...         
5765            JONES, Jenny
5766         ANDERSON, Jamie
5767      MALTAIS, Dominique
5768            SAMKOVA, Eva
5769        TRESPEUCH, Chloe
Name: Athlete Name, Length: 5770, dtype: object

### If you pass a list of multiple column headers, it returns dataframe with only those columns

####  ```df[["Column 1", "Column 2"]]```

In [23]:
## Create a dataframe that for only summer games years and cities
dfs_Year_City = dfs[["Year", "City"]]
dfs_Year_City

Year    City
0      1896  Athens
1      1896  Athens
2      1896  Athens
3      1896  Athens
4      1896  Athens
...     ...     ...
31160  2012  London
31161  2012  London
31162  2012  London
31163  2012  London
31164  2012  London

[31165 rows x 2 columns]

In [24]:
dfs_Year_City.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31165 entries, 0 to 31164
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Year    31165 non-null  int64 
 1   City    31165 non-null  object
dtypes: int64(1), object(1)
memory usage: 487.1+ KB


## Sorting by single value

```## are the two column headers identical?
list(dfs.columns) == list(dfw.columns))``` for A-Z or smaller numbers to bigger numbers

```ascending = False``` for Z-A or bigger to smaller numbers.

In [27]:
## sort by athlete names A-Z for winter olympics
dfw.sort_values (by = "Athlete Name", ascending = True)

Year          City       Sport            Discipline  \
1834  1980   Lake Placid  Ice Hockey            Ice Hockey   
2052  1984      Sarajevo  Ice Hockey            Ice Hockey   
1985  1980   Lake Placid      Skiing  Cross Country Skiing   
5303  2014         Sochi  Ice Hockey            Ice Hockey   
2738  1992   Albertville      Skiing         Alpine Skiing   
...    ...           ...         ...                   ...   
2920  1994   Lillehammer  Ice Hockey            Ice Hockey   
994   1960  Squaw Valley      Skiing  Cross Country Skiing   
549   1948     St.Moritz      Skiing  Cross Country Skiing   
556   1948     St.Moritz      Skiing  Cross Country Skiing   
1704  1976     Innsbruck     Skating        Figure skating   

              Athlete Name Athlete Country Gender         Event   Medal  
1834        AAHLBERG, Mats             SWE    Men    Ice Hockey  Bronze  
2052        AAHLEN, Thomas             SWE    Men    Ice Hockey  Bronze  
1985      AALAND, Per Knut             NOR    Men  4X10KM Relay  Silver  
5303   AALTONEN, Juhamatti             FIN    Men    Ice Hockey  Bronze  
2738  AAMODT, Kjetil Andre             NOR    Men       Super-G    Gold  
...                    ...             ...    ...           ...     ...  
2920       ÖRNSKOG, Stefan             SWE    Men    Ice Hockey    Gold  
994           ÖSTBY, Einar             NOR    Men  4X10KM Relay  Silver  
549        ÖSTENSSON, Nils             SWE    Men          18KM  Silver  
556        ÖSTENSSON, Nils             SWE    Men  4X10KM Relay    Gold  
1704      ÖSTERREICH, Rolf             GDR    Men         Pairs  Silver  

[5770 rows x 9 columns]

## Sorting by multiple values

```dataframe.sort_values(by=["Value 1", "Value 2"], ascending = True)``` for A-Z or smaller numbers to bigger numbers

```ascending = False``` for Z-A or bigger to smaller numbers.


In [28]:
## Sort by values (City and Year) for winter olympics
dfw.sort_values(by = ["City", "Year"], ascending = True)

Year         City     Sport Discipline        Athlete Name  \
2502  1992  Albertville  Biathlon   Biathlon     ELORANTA, Harri   
2503  1992  Albertville  Biathlon   Biathlon      KIRCHNER, Mark   
2504  1992  Albertville  Biathlon   Biathlon        GROSS, Ricco   
2505  1992  Albertville  Biathlon   Biathlon      BEDARD, Myriam   
2506  1992  Albertville  Biathlon   Biathlon       HARVEY, Antje   
...    ...          ...       ...        ...                 ...   
5153  2010    Vancouver    Skiing  Snowboard       WESCOTT, Seth   
5154  2010    Vancouver    Skiing  Snowboard     ROBERTSON, Mike   
5155  2010    Vancouver    Skiing  Snowboard        NOBS, Olivia   
5156  2010    Vancouver    Skiing  Snowboard      RICKER, Maelle   
5157  2010    Vancouver    Skiing  Snowboard  ANTHONIOZ, Deborah   

     Athlete Country Gender            Event   Medal  
2502             FIN    Men             10KM  Bronze  
2503             GER    Men             10KM    Gold  
2504             GER    Men             10KM  Silver  
2505             CAN  Women             15KM  Bronze  
2506             GER  Women             15KM    Gold  
...              ...    ...              ...     ...  
5153             USA    Men  Snowboard Cross    Gold  
5154             CAN    Men  Snowboard Cross  Silver  
5155             SUI  Women  Snowboard Cross  Bronze  
5156             CAN  Women  Snowboard Cross    Gold  
5157             FRA  Women  Snowboard Cross  Silver  

[5770 rows x 9 columns]

In [29]:
## sort by city and year but descending for summer games
dfs.sort_values(by=["City", "Year"], ascending = False)

Year       City      Sport       Discipline             Athlete Name  \
10674  1964      Tokyo   Aquatics           Diving      GOMPF, Thomas Eugen   
10675  1964      Tokyo   Aquatics           Diving    WEBSTER, Robert David   
10676  1964      Tokyo   Aquatics           Diving           DIBIASI, Klaus   
10677  1964      Tokyo   Aquatics           Diving        ALEKSEEVA, Galina   
10678  1964      Tokyo   Aquatics           Diving       BUSH, Lesley Leigh   
...     ...        ...        ...              ...                      ...   
5709   1928  Amsterdam  Wrestling  Wrestling Gre-R  KOKKINEN, Väinö Anselmi   
5710   1928  Amsterdam  Wrestling  Wrestling Gre-R                 PAPP, L.   
5711   1928  Amsterdam  Wrestling  Wrestling Gre-R           PELLINEN, Onni   
5712   1928  Amsterdam  Wrestling  Wrestling Gre-R       MOUSTAPHA, Ibrahim   
5713   1928  Amsterdam  Wrestling  Wrestling Gre-R               RIEGER, A.   

      Athlete Country Gender                            Event   Medal  
10674             USA    Men                     10M Platform  Bronze  
10675             USA    Men                     10M Platform    Gold  
10676             ITA    Men                     10M Platform  Silver  
10677             URS  Women                     10M Platform  Bronze  
10678             USA  Women                     10M Platform    Gold  
...               ...    ...                              ...     ...  
5709              FIN    Men       67.5 - 75KG (Middleweight)    Gold  
5710              HUN    Men       67.5 - 75KG (Middleweight)  Silver  
5711              FIN    Men  75 - 82.5KG (Light-Heavyweight)  Bronze  
5712              EGY    Men  75 - 82.5KG (Light-Heavyweight)    Gold  
5713              GER    Men  75 - 82.5KG (Light-Heavyweight)  Silver  

[31165 rows x 9 columns]

In [30]:
## you can store into a new df
dfx = dfs.sort_values(by=["City", "Year"], ascending = False)

## Should Winter and Summer data remain separate?


### Creating a new column with a default value

In [32]:
## In the summer dataframe, add "Season" column with the value "Summer"
dfs ["Season" ] = "Summer"
dfs

Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                    Athlete Name Athlete Country Gender  \
0                  HAJOS, Alfred             HUN    Men   
1               HERSCHMANN, Otto             AUT    Men   
2              DRIVAS, Dimitrios             GRE    Men   
3             MALOKINIS, Ioannis             GRE    Men   
4             CHASAPIS, Spiridon             GRE    Men   
...                          ...             ...    ...   
31160         JANIKOWSKI, Damian             POL    Men   
31161  REZAEI, Ghasem Gholamreza             IRI    Men   
31162             TOTROV, Rustam             RUS    Men   
31163          ALEKSANYAN, Artur             ARM    Men   
31164             LIDBERG, Jimmy             SWE    Men   

                            Event   Medal  Season  
0                  100M Freestyle    Gold  Summer  
1                  100M Freestyle  Silver  Summer  
2      100M Freestyle For Sailors  Bronze  Summer  
3      100M Freestyle For Sailors    Gold  Summer  
4      100M Freestyle For Sailors  Silver  Summer  
...                           ...     ...     ...  
31160                    Wg 84 KG  Bronze  Summer  
31161                    Wg 96 KG    Gold  Summer  
31162                    Wg 96 KG  Silver  Summer  
31163                    Wg 96 KG  Bronze  Summer  
31164                    Wg 96 KG  Bronze  Summer  

[31165 rows x 10 columns]

In [33]:
## In the winter dataframe, add "Season" column with the value "Winter"
dfw ["Season" ] = "Winter"
dfw

Year      City     Sport Discipline          Athlete Name  \
0     1924  Chamonix  Biathlon   Biathlon           BERTHET, G.   
1     1924  Chamonix  Biathlon   Biathlon        MANDRILLON, C.   
2     1924  Chamonix  Biathlon   Biathlon   MANDRILLON, Maurice   
3     1924  Chamonix  Biathlon   Biathlon       VANDELLE, André   
4     1924  Chamonix  Biathlon   Biathlon  AUFDENBLATTEN, Adolf   
...    ...       ...       ...        ...                   ...   
5765  2014     Sochi    Skiing  Snowboard          JONES, Jenny   
5766  2014     Sochi    Skiing  Snowboard       ANDERSON, Jamie   
5767  2014     Sochi    Skiing  Snowboard    MALTAIS, Dominique   
5768  2014     Sochi    Skiing  Snowboard          SAMKOVA, Eva   
5769  2014     Sochi    Skiing  Snowboard      TRESPEUCH, Chloe   

     Athlete Country Gender            Event   Medal  Season  
0                FRA    Men  Military Patrol  Bronze  Winter  
1                FRA    Men  Military Patrol  Bronze  Winter  
2                FRA    Men  Military Patrol  Bronze  Winter  
3                FRA    Men  Military Patrol  Bronze  Winter  
4                SUI    Men  Military Patrol    Gold  Winter  
...              ...    ...              ...     ...     ...  
5765             GBR  Women       Slopestyle  Bronze  Winter  
5766             USA  Women       Slopestyle    Gold  Winter  
5767             CAN  Women  Snowboard Cross  Silver  Winter  
5768             CZE  Women  Snowboard Cross    Gold  Winter  
5769             FRA  Women  Snowboard Cross  Bronze  Winter  

[5770 rows x 10 columns]

In [34]:
## are the two column headers identical?
list(dfs.columns) == list(dfw.columns)

True

In [38]:
## join winter and summer dataframes together into a dataframe called oly
oly = pd.concat ([dfs,dfw], ignore_index = True) # ignore index es para que no mantenga esa columna y haga una nueva

In [39]:
oly

Year    City     Sport Discipline        Athlete Name Athlete Country  \
0      1896  Athens  Aquatics   Swimming       HAJOS, Alfred             HUN   
1      1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto             AUT   
2      1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios             GRE   
3      1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis             GRE   
4      1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon             GRE   
...     ...     ...       ...        ...                 ...             ...   
36930  2014   Sochi    Skiing  Snowboard        JONES, Jenny             GBR   
36931  2014   Sochi    Skiing  Snowboard     ANDERSON, Jamie             USA   
36932  2014   Sochi    Skiing  Snowboard  MALTAIS, Dominique             CAN   
36933  2014   Sochi    Skiing  Snowboard        SAMKOVA, Eva             CZE   
36934  2014   Sochi    Skiing  Snowboard    TRESPEUCH, Chloe             FRA   

      Gender                       Event   Medal  Season  
0        Men              100M Freestyle    Gold  Summer  
1        Men              100M Freestyle  Silver  Summer  
2        Men  100M Freestyle For Sailors  Bronze  Summer  
3        Men  100M Freestyle For Sailors    Gold  Summer  
4        Men  100M Freestyle For Sailors  Silver  Summer  
...      ...                         ...     ...     ...  
36930  Women                  Slopestyle  Bronze  Winter  
36931  Women                  Slopestyle    Gold  Winter  
36932  Women             Snowboard Cross  Silver  Winter  
36933  Women             Snowboard Cross    Gold  Winter  
36934  Women             Snowboard Cross  Bronze  Winter  

[36935 rows x 10 columns]

In [41]:
### confirm that you have the correct number of rows after the join.
### you can scroll up to count the number of entries in oly_w and oly_s. 
### The totals should be equal to the value generated below
dfw.shape[0] + dfs.shape[0] == oly.shape[0]


True

In [42]:
## show top and bottom
oly

Year    City     Sport Discipline        Athlete Name Athlete Country  \
0      1896  Athens  Aquatics   Swimming       HAJOS, Alfred             HUN   
1      1896  Athens  Aquatics   Swimming    HERSCHMANN, Otto             AUT   
2      1896  Athens  Aquatics   Swimming   DRIVAS, Dimitrios             GRE   
3      1896  Athens  Aquatics   Swimming  MALOKINIS, Ioannis             GRE   
4      1896  Athens  Aquatics   Swimming  CHASAPIS, Spiridon             GRE   
...     ...     ...       ...        ...                 ...             ...   
36930  2014   Sochi    Skiing  Snowboard        JONES, Jenny             GBR   
36931  2014   Sochi    Skiing  Snowboard     ANDERSON, Jamie             USA   
36932  2014   Sochi    Skiing  Snowboard  MALTAIS, Dominique             CAN   
36933  2014   Sochi    Skiing  Snowboard        SAMKOVA, Eva             CZE   
36934  2014   Sochi    Skiing  Snowboard    TRESPEUCH, Chloe             FRA   

      Gender                       Event   Medal  Season  
0        Men              100M Freestyle    Gold  Summer  
1        Men              100M Freestyle  Silver  Summer  
2        Men  100M Freestyle For Sailors  Bronze  Summer  
3        Men  100M Freestyle For Sailors    Gold  Summer  
4        Men  100M Freestyle For Sailors  Silver  Summer  
...      ...                         ...     ...     ...  
36930  Women                  Slopestyle  Bronze  Winter  
36931  Women                  Slopestyle    Gold  Winter  
36932  Women             Snowboard Cross  Silver  Winter  
36933  Women             Snowboard Cross    Gold  Winter  
36934  Women             Snowboard Cross  Bronze  Winter  

[36935 rows x 10 columns]

In [43]:
### generate a random sample of 20 rows
### if you see only "summer" or only "winter" in the "Season" column, just run the cell again to
## confirm both seasons are in oly.
oly.sample(20)


Year                    City          Sport            Discipline  \
18631  1988                   Seoul        Cycling          Cycling Road   
13970  1976                Montreal       Aquatics              Swimming   
23129  1996                 Atlanta      Wrestling       Wrestling Free.   
26302  2004                  Athens     Gymnastics           Artistic G.   
18926  1988                   Seoul     Gymnastics           Artistic G.   
12602  1968                  Mexico     Volleyball            Volleyball   
36926  2014                   Sochi         Skiing             Snowboard   
21939  1996                 Atlanta         Boxing                Boxing   
33370  1984                Sarajevo         Skiing  Cross Country Skiing   
907    1904                St Louis       Football              Football   
31594  1936  Garmisch Partenkirchen         Skiing           Ski Jumping   
2804   1912               Stockholm     Tug of War            Tug of War   
18596  1988                   Seoul  Canoe / Kayak       Canoe / Kayak F   
33250  1984                Sarajevo     Ice Hockey            Ice Hockey   
25596  2004                  Athens      Athletics             Athletics   
29094  2008                 Beijing     Volleyball            Volleyball   
18587  1988                   Seoul  Canoe / Kayak       Canoe / Kayak F   
19223  1988                   Seoul         Rowing                Rowing   
15790  1980                  Moscow     Equestrian               Jumping   
18406  1988                   Seoul      Athletics             Athletics   

                       Athlete Name Athlete Country Gender  \
18631                  GRONE, Bernd             FRG    Men   
13970              STRACHAN, Rodney             USA    Men   
23129                   ANGLE, Kurt             USA    Men   
26302               YANG, Tae Young             KOR    Men   
18926             POTORAC, Gabriela             ROU  Women   
12602            MITSUMORI, Yasuaki             JPN    Men   
36926              DUJMOVITS, Julia             AUT  Women   
21939                 MESA, Arnaldo             CUB    Men   
33370            SVAN, Gunde Anders             SWE    Men   
907        JANUARY, Thomas Thurston             USA    Men   
31594                  RUUD, Birger             NOR    Men   
2804               MUNRO, Alexander             GBR    Men   
18596  MAC DONALD, Christopher Paul             NZL    Men   
33250            PERVUKHINE, Vasili             URS    Men   
25596              ARRON, Christine             FRA  Women   
29094              JOINES, Jennifer             USA  Women   
18587                 FOSTER, Peter             AUS    Men   
19223                GRABOW, Volker             FRG    Men   
15790                    LADY MIRKA             MEX  Women   
18406                CONWAY, Hollis             USA    Men   

                            Event   Medal  Season  
18631        Individual Road Race  Silver  Summer  
13970      400M Individual Medley    Gold  Summer  
23129    90 - 100KG (Heavyweight)    Gold  Summer  
26302        Individual All-Round  Bronze  Summer  
18926                       Vault  Silver  Summer  
12602                  Volleyball  Silver  Summer  
36926             Parallel Slalom    Gold  Winter  
21939    51 - 54KG (Bantamweight)  Silver  Summer  
33370                4X10KM Relay    Gold  Winter  
907                      Football    Gold  Summer  
31594        K90 Individual (70M)    Gold  Winter  
2804                   Tug Of War  Silver  Summer  
18596     K-2 500M (Kayak Double)    Gold  Summer  
33250                  Ice Hockey    Gold  Winter  
25596                4X100M Relay  Bronze  Summer  
29094                  Volleyball  Silver  Summer  
18587    K-2 1000M (Kayak Double)  Bronze  Summer  
19223  Four Without Coxswain (4-)  Bronze  Summer  
15790                        Team  Bronze  Summer  
18406                   High Jump  Silver  Summer

In [45]:
oly.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36935 entries, 0 to 36934
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Year             36935 non-null  int64 
 1   City             36935 non-null  object
 2   Sport            36935 non-null  object
 3   Discipline       36935 non-null  object
 4   Athlete Name     36935 non-null  object
 5   Athlete Country  36931 non-null  object
 6   Gender           36935 non-null  object
 7   Event            36935 non-null  object
 8   Medal            36935 non-null  object
 9   Season           36935 non-null  object
dtypes: int64(1), object(9)
memory usage: 2.8+ MB


## Improving memory allocation

If we were working with a massive dataset, we'd want to find ways to improve processing power.

For example, the following columns have only a few values that repeat again and again as strings. This is highly inefficient. 

- The "Season" column has "winter" and "summer", 
- The "Medal" column only has "gold", "silver" and "bronze",
- "Gender" at this point in the Olympics only has "male" and "female".

One way to improve memory allocation is to take columns with that contain only a few data points and turn them into categories. 



In [48]:
## let's first get info() on oly.
## you should see how "Season", "Medal" and "Gender" are all string objects.
## Note that memory usage: 2.8+ MB
# Lo vamos a convertir en categorias (como mucho, unas cinco categorias)
oly [["Medal", "Gender", "Season"]] = \
oly [["Medal", "Gender", "Season"]].astype("category")

In [49]:
oly [["Medal", "Gender", "Season"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36935 entries, 0 to 36934
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   Medal   36935 non-null  category
 1   Gender  36935 non-null  category
 2   Season  36935 non-null  category
dtypes: category(3)
memory usage: 108.7 KB


# Filter/Subset 



### Create a subset that holds the results from the Sochi Olympics

**Old way:**

```df[df["column_name"] == "value_in_column"]```

**Newer way:**

```df.query("ColumnName == 'some_value'")```

In [51]:
## return df that holds only Sochi data

df_sochi = oly.query("City == 'Sochi'")
df_sochi

Year   City     Sport Discipline             Athlete Name  \
36323  2014  Sochi  Biathlon   Biathlon    LANDERTINGER, Dominik   
36324  2014  Sochi  Biathlon   Biathlon         SOUKUP, Jaroslav   
36325  2014  Sochi  Biathlon   Biathlon   BJOERNDALEN, Ole Einar   
36326  2014  Sochi  Biathlon   Biathlon          MORAVEC, Ondrej   
36327  2014  Sochi  Biathlon   Biathlon  BEATRIX, Jean Guillaume   
...     ...    ...       ...        ...                      ...   
36930  2014  Sochi    Skiing  Snowboard             JONES, Jenny   
36931  2014  Sochi    Skiing  Snowboard          ANDERSON, Jamie   
36932  2014  Sochi    Skiing  Snowboard       MALTAIS, Dominique   
36933  2014  Sochi    Skiing  Snowboard             SAMKOVA, Eva   
36934  2014  Sochi    Skiing  Snowboard         TRESPEUCH, Chloe   

      Athlete Country Gender            Event   Medal  Season  
36323             AUT    Men             10KM  Silver  Winter  
36324             CZE    Men             10KM  Bronze  Winter  
36325             NOR    Men             10KM    Gold  Winter  
36326             CZE    Men   12.5Km Pursuit  Silver  Winter  
36327             FRA    Men   12.5Km Pursuit  Bronze  Winter  
...               ...    ...              ...     ...     ...  
36930             GBR  Women       Slopestyle  Bronze  Winter  
36931             USA  Women       Slopestyle    Gold  Winter  
36932             CAN  Women  Snowboard Cross  Silver  Winter  
36933             CZE  Women  Snowboard Cross    Gold  Winter  
36934             FRA  Women  Snowboard Cross  Bronze  Winter  

[612 rows x 10 columns]

**Syntax for querying multi-word column names:**

```dataframe.query('`column name` comparison_operator value')```

Note the opening and closing tick  ``` ` ``` marks.

In [54]:
## query to create df so that only athletes from France appear
df_france = oly.query ("`Athlete Country` == 'FRA'")
df_france

Year    City      Sport        Discipline        Athlete Name  \
17     1896  Athens  Athletics         Athletics   LERMUSIAUX, Albin   
47     1896  Athens  Athletics         Athletics  TUFFERI, Alexandre   
51     1896  Athens    Cycling     Cycling Track       FLAMENG, Léon   
54     1896  Athens    Cycling     Cycling Track        MASSON, Paul   
55     1896  Athens    Cycling     Cycling Track       FLAMENG, Léon   
...     ...     ...        ...               ...                 ...   
36846  2014   Sochi     Skiing  Freestyle Skiing      ROLLAND, Kevin   
36860  2014   Sochi     Skiing  Freestyle Skiing     MARTINOD, Marie   
36903  2014   Sochi     Skiing       Ski Jumping      MATTEL, Coline   
36917  2014   Sochi     Skiing         Snowboard    VAULTIER, Pierre   
36934  2014   Sochi     Skiing         Snowboard    TRESPEUCH, Chloe   

      Athlete Country Gender            Event   Medal  Season  
17                FRA    Men            1500M  Bronze  Summer  
47                FRA    Men      Triple Jump  Silver  Summer  
51                FRA    Men            100KM    Gold  Summer  
54                FRA    Men             10KM    Gold  Summer  
55                FRA    Men             10KM  Silver  Summer  
...               ...    ...              ...     ...     ...  
36846             FRA    Men     Ski Halfpipe  Bronze  Winter  
36860             FRA  Women     Ski Halfpipe  Silver  Winter  
36903             FRA  Women   K90 Individual  Bronze  Winter  
36917             FRA    Men  Snowboard Cross    Gold  Winter  
36934             FRA  Women  Snowboard Cross  Bronze  Winter  

[1548 rows x 10 columns]

## How many values in a column?

```df_name["column_name"].value_counts()```

In [57]:
## How many men and women won medals Sochi?
oly["Gender"].value_counts()

Men      26690
Women    10245
Name: Gender, dtype: int64

## Filtering by dates

You'll learn more about date and time in the coming week. For now, please note how the ```Year``` column is a  ```int64``` object and NOT a ```datetime``` object. We can still work with it for our needs.

In [60]:
## Find all the 1900 competitions
df_1900 = oly.query ("Year == 1900")
df_1900

Year   City       Sport  Discipline                       Athlete Name  \
151  1900  Paris    Aquatics    Swimming                     HALMAY, Zoltan   
152  1900  Paris    Aquatics    Swimming                JARVIS, John Arthur   
153  1900  Paris    Aquatics    Swimming                        WAHLE, Otto   
154  1900  Paris    Aquatics    Swimming                    DROST, Johannes   
155  1900  Paris    Aquatics    Swimming                  HOPPENBERG, Ernst   
..    ...    ...         ...         ...                                ...   
658  1900  Paris  Tug of War  Tug of War                       COLLAS, Jean   
659  1900  Paris  Tug of War  Tug of War                  GONDOUIN, Charles   
660  1900  Paris  Tug of War  Tug of War  HENRIQUEZ DE ZUBIERRA, Constantin   
661  1900  Paris  Tug of War  Tug of War                      ROFFO, Joseph   
662  1900  Paris  Tug of War  Tug of War                     SARRADE, Emile   

    Athlete Country Gender            Event   Medal  Season  
151             HUN    Men  1500M Freestyle  Bronze  Summer  
152             GBR    Men  1500M Freestyle    Gold  Summer  
153             AUT    Men  1500M Freestyle  Silver  Summer  
154             NED    Men  200M Backstroke  Bronze  Summer  
155             GER    Men  200M Backstroke    Gold  Summer  
..              ...    ...              ...     ...     ...  
658             FRA    Men       Tug Of War  Silver  Summer  
659             FRA    Men       Tug Of War  Silver  Summer  
660             FRA    Men       Tug Of War  Silver  Summer  
661             FRA    Men       Tug Of War  Silver  Summer  
662             FRA    Men       Tug Of War  Silver  Summer  

[512 rows x 10 columns]

In [63]:
### return all the competitions between 1920 and 1950.
df_1920_1950 = oly.query ("1920 <= Year <=1950")
df_1920_1950

Year       City     Sport       Discipline              Athlete Name  \
2822   1920    Antwerp  Aquatics           Diving            PRIESTE, Harry   
2823   1920    Antwerp  Aquatics           Diving        PINKSTON, Clarence   
2824   1920    Antwerp  Aquatics           Diving              ADLERZ, Erik   
2825   1920    Antwerp  Aquatics           Diving             OLLIVIER, Eva   
2826   1920    Antwerp  Aquatics           Diving  FRYLAND CLAUSEN, Stefani   
...     ...        ...       ...              ...                       ...   
31731  1948  St.Moritz    Skiing  Nordic Combined              HASU, Heikki   
31732  1948  St.Moritz    Skiing  Nordic Combined           HUHTALA, Martti   
31733  1948  St.Moritz    Skiing      Ski Jumping     SCHJELDERUP, Thorleif   
31734  1948  St.Moritz    Skiing      Ski Jumping           HUGSTED, Petter   
31735  1948  St.Moritz    Skiing      Ski Jumping              RUUD, Birger   

      Athlete Country Gender                 Event   Medal  Season  
2822              USA    Men          10M Platform  Bronze  Summer  
2823              USA    Men          10M Platform    Gold  Summer  
2824              SWE    Men          10M Platform  Silver  Summer  
2825              SWE  Women          10M Platform  Bronze  Summer  
2826              DEN  Women          10M Platform    Gold  Summer  
...               ...    ...                   ...     ...     ...  
31731             FIN    Men            Individual    Gold  Winter  
31732             FIN    Men            Individual  Silver  Winter  
31733             NOR    Men  K90 Individual (70M)  Bronze  Winter  
31734             NOR    Men  K90 Individual (70M)    Gold  Winter  
31735             NOR    Men  K90 Individual (70M)  Silver  Winter  

[5767 rows x 10 columns]

In [67]:
## return only Tennis competitions between 1920 and 1950.
df_1920_1950_Tennis = oly.query ("1920 <= Year <=1950 & Sport == 'Tennis'")
df_1920_1950_Tennis

Year     City   Sport Discipline                       Athlete Name  \
4026  1920  Antwerp  Tennis     Tennis                   ALBARRAN, Pierre   
4027  1920  Antwerp  Tennis     Tennis                       DECUGIS, Max   
4028  1920  Antwerp  Tennis     Tennis       TURNBULL, Oswald Graham Noel   
4029  1920  Antwerp  Tennis     Tennis                   WOOSNAM, Maxwell   
4030  1920  Antwerp  Tennis     Tennis                  KASHIO, Seiichiro   
4031  1920  Antwerp  Tennis     Tennis                    KUMAGAE, Ichiya   
4032  1920  Antwerp  Tennis     Tennis                  D'AYEN, Elisabeth   
4033  1920  Antwerp  Tennis     Tennis                   LENGLEN, Suzanne   
4034  1920  Antwerp  Tennis     Tennis                   MCKANE, Kathleen   
4035  1920  Antwerp  Tennis     Tennis          MCNAIR, Winifred Margaret   
4036  1920  Antwerp  Tennis     Tennis        BEAMISH, Winifred Geraldine   
4037  1920  Antwerp  Tennis     Tennis              HOLMAN, Edith Dorothy   
4038  1920  Antwerp  Tennis     Tennis                    ZEMLA, Ladislav   
4039  1920  Antwerp  Tennis     Tennis                       DECUGIS, Max   
4040  1920  Antwerp  Tennis     Tennis                   WOOSNAM, Maxwell   
4041  1920  Antwerp  Tennis     Tennis                   SKRBKOVA, Milada   
4042  1920  Antwerp  Tennis     Tennis                   LENGLEN, Suzanne   
4043  1920  Antwerp  Tennis     Tennis                   MCKANE, Kathleen   
4044  1920  Antwerp  Tennis     Tennis         WINSLOW, Charles Lyndhurst   
4045  1920  Antwerp  Tennis     Tennis                     RAYMOND, Louis   
4046  1920  Antwerp  Tennis     Tennis                    KUMAGAE, Ichiya   
4047  1920  Antwerp  Tennis     Tennis                   MCKANE, Kathleen   
4048  1920  Antwerp  Tennis     Tennis                   LENGLEN, Suzanne   
4049  1920  Antwerp  Tennis     Tennis              HOLMAN, Edith Dorothy   
4926  1924    Paris  Tennis     Tennis                   BOROTRA, Jean R.   
4927  1924    Paris  Tennis     Tennis                 LACOSTE, Jean René   
4928  1924    Paris  Tennis     Tennis           HUNTER, Francis Townsend   
4929  1924    Paris  Tennis     Tennis                  RICHARDS, Vincent   
4930  1924    Paris  Tennis     Tennis                   BRUGNON, Jacques   
4931  1924    Paris  Tennis     Tennis                      COCHET, Henri   
4932  1924    Paris  Tennis     Tennis                COLYER, Evelyn Lucy   
4933  1924    Paris  Tennis     Tennis  SHEPHERD BARRON, Dorothy Cunliffe   
4934  1924    Paris  Tennis     Tennis           WIGHTMAN, Hazel Virginia   
4935  1924    Paris  Tennis     Tennis                 WILLS-MOODY, Helen   
4936  1924    Paris  Tennis     Tennis            COVELL, Phyllis Lindrea   
4937  1924    Paris  Tennis     Tennis                 MAC KANE, Kathleen   
4938  1924    Paris  Tennis     Tennis                    TIMMER, Hendrik   
4939  1924    Paris  Tennis     Tennis       WILLIAMS, Richard Norris III   
4940  1924    Paris  Tennis     Tennis                  RICHARDS, Vincent   
4941  1924    Paris  Tennis     Tennis                   BOUMAN, Cornelia   
4942  1924    Paris  Tennis     Tennis           WIGHTMAN, Hazel Virginia   
4943  1924    Paris  Tennis     Tennis                     JESSUP, Marion   
4944  1924    Paris  Tennis     Tennis               DE MORPURGO, Umberto   
4945  1924    Paris  Tennis     Tennis                  RICHARDS, Vincent   
4946  1924    Paris  Tennis     Tennis                      COCHET, Henri   
4947  1924    Paris  Tennis     Tennis                 MAC KANE, Kathleen   
4948  1924    Paris  Tennis     Tennis                 WILLS-MOODY, Helen   
4949  1924    Paris  Tennis     Tennis                   VLASTO, Julie P.   

     Athlete Country Gender          Event   Medal  Season  
4026             FRA    Men        Doubles  Bronze  Summer  
4027             FRA    Men        Doubles  Bronze  Summer  
4028             GBR    Men        Doubles    Gol

In [70]:
## return only Women's Tennis competitions between 1996 and 2019 
df_1966_2019_Tennis_W = oly.query ("1996 <= Year <= 2019 \
                                    & Sport == 'Tennis' \
                                    & Gender == 'Women'")
df_1966_2019_Tennis_W                      

Year     City   Sport Discipline              Athlete Name  \
22975  1996  Atlanta  Tennis     Tennis        MARTINEZ, Conchita   
22976  1996  Atlanta  Tennis     Tennis  SANCHEZ-VICARIO, Arantxa   
22977  1996  Atlanta  Tennis     Tennis           FERNANDEZ, Gigi   
22978  1996  Atlanta  Tennis     Tennis       FERNANDEZ, Mary Joe   
22979  1996  Atlanta  Tennis     Tennis             NOVOTNA, Jana   
22980  1996  Atlanta  Tennis     Tennis            SUKOVA, Helena   
22984  1996  Atlanta  Tennis     Tennis             NOVOTNA, Jana   
22985  1996  Atlanta  Tennis     Tennis        DAVENPORT, Lindsay   
22986  1996  Atlanta  Tennis     Tennis  SANCHEZ-VICARIO, Arantxa   
24981  2000   Sydney  Tennis     Tennis              CALLENS, Els   
24982  2000   Sydney  Tennis     Tennis      VAN ROOST, Dominique   
24983  2000   Sydney  Tennis     Tennis          WILLIAMS, Serena   
24984  2000   Sydney  Tennis     Tennis           WILLIAMS, Venus   
24985  2000   Sydney  Tennis     Tennis          BOOGERT, Kristie   
24986  2000   Sydney  Tennis     Tennis           OREMANS, Miriam   
24990  2000   Sydney  Tennis     Tennis             SELES, Monica   
24991  2000   Sydney  Tennis     Tennis           WILLIAMS, Venus   
24992  2000   Sydney  Tennis     Tennis         DEMENTIEVA, Elena   
26976  2004   Athens  Tennis     Tennis             SUAREZ, Paola   
26977  2004   Athens  Tennis     Tennis        TARABINI, Patricia   
26978  2004   Athens  Tennis     Tennis                  LI, Ting   
26979  2004   Athens  Tennis     Tennis            SUN, Tian Tian   
26983  2004   Athens  Tennis     Tennis             MOLIK, Alicia   
26984  2004   Athens  Tennis     Tennis   HENIN-HARDENNE, Justine   
26985  2004   Athens  Tennis     Tennis          MAURESMO, Amelie   
28998  2008  Beijing  Tennis     Tennis                   YAN, Zi   
28999  2008  Beijing  Tennis     Tennis                ZHENG, Jie   
29000  2008  Beijing  Tennis     Tennis          WILLIAMS, Serena   
29001  2008  Beijing  Tennis     Tennis           WILLIAMS, Venus   
29002  2008  Beijing  Tennis     Tennis  MEDINA GARRIGUES, Anabel   
29003  2008  Beijing  Tennis     Tennis   RUANO PASCUAL, Virginia   
29007  2008  Beijing  Tennis     Tennis           ZVONAREVA, Vera   
29008  2008  Beijing  Tennis     Tennis         DEMENTIEVA, Elena   
29009  2008  Beijing  Tennis     Tennis            SAFINA, Dinara   
30940  2012   London  Tennis     Tennis          WILLIAMS, Serena   
30941  2012   London  Tennis     Tennis           WILLIAMS, Venus   
30942  2012   London  Tennis     Tennis        HLAVACKOVA, Andrea   
30943  2012   London  Tennis     Tennis           HRADECKA, Lucie   
30944  2012   London  Tennis     Tennis          KIRILENKO, Maria   
30945  2012   London  Tennis     Tennis            PETROVA, Nadia   
30946  2012   London  Tennis     Tennis        AZARENKA, Victoria   
30949  2012   London  Tennis     Tennis             ROBSON, Laura   
30951  2012   London  Tennis     Tennis             RAYMOND, Lisa   
30955  2012   London  Tennis     Tennis          WILLIAMS, Serena   
30956  2012   London  Tennis     Tennis          SHARAPOVA, Maria   
30957  2012   London  Tennis     Tennis        AZARENKA, Victoria   

      Athlete Country Gender          Event   Medal  Season  
22975             ESP  Women        Doubles  Bronze  Summer  
22976             ESP  Women        Doubles  Bronze  Summer  
22977             USA  Women        Doubles    Gold  Summer  
22978             USA  Women        Doubles    Gold  Summer  
22979             CZE  Women        Doubles  Silver  Summer  
22980             CZE  Women        Doubles  Silver  Summer  
22984             CZE  Women        Singles  Bronze  Summer  
22985             USA  Women        Singles    Gold  Summer  
22986             ESP  Women        Singles  Silver  Summer  
24981             BEL  Women        Doubles  Bronze  Summer  
24982             BEL  Women        Doubles  Bronze  Summer  
24983             U

In [80]:
## Mujeres jugando tenis o box
df_Tennis_Box_W = oly.query ("Gender == 'Women' & (Sport == 'Tennis' | Sport == 'Boxing')")
df_Tennis_Box_W

Year    City   Sport Discipline        Athlete Name Athlete Country  \
639    1900   Paris  Tennis     Tennis       JONES, Marion             ZZX   
640    1900   Paris  Tennis     Tennis   ROSENBAUM, Hedwig             ZZX   
641    1900   Paris  Tennis     Tennis   COOPER, Charlotte             GBR   
642    1900   Paris  Tennis     Tennis     PREVOST, Hélène             ZZX   
647    1900   Paris  Tennis     Tennis       JONES, Marion             USA   
...     ...     ...     ...        ...                 ...             ...   
30949  2012  London  Tennis     Tennis       ROBSON, Laura             GBR   
30951  2012  London  Tennis     Tennis       RAYMOND, Lisa             USA   
30955  2012  London  Tennis     Tennis    WILLIAMS, Serena             USA   
30956  2012  London  Tennis     Tennis    SHARAPOVA, Maria             RUS   
30957  2012  London  Tennis     Tennis  AZARENKA, Victoria             BLR   

      Gender          Event   Medal  Season  
639    Women  Mixed Doubles  Bronze  Summer  
640    Women  Mixed Doubles  Bronze  Summer  
641    Women  Mixed Doubles    Gold  Summer  
642    Women  Mixed Doubles  Silver  Summer  
647    Women        Singles  Bronze  Summer  
...      ...            ...     ...     ...  
30949  Women  Mixed Doubles  Silver  Summer  
30951  Women  Mixed Doubles  Bronze  Summer  
30955  Women        Singles    Gold  Summer  
30956  Women        Singles  Silver  Summer  
30957  Women        Singles  Bronze  Summer  

[132 rows x 10 columns]

## What year was women's boxing introduced to the olympics?

In [82]:
## find all the names of sports first just to confirm boxing 
## does not appear in different names
## how is boxing listed?

In [85]:
oly["Sport"].unique()

array(['Aquatics', 'Athletics', 'Cycling', 'Fencing', 'Gymnastics',
       'Shooting', 'Tennis', 'Weightlifting', 'Wrestling', 'Archery',
       'Basque Pelota', 'Cricket', 'Croquet', 'Equestrian', 'Football',
       'Golf', 'Polo', 'Rowing', 'Rugby', 'Sailing', 'Tug of War',
       'Boxing', 'Lacrosse', 'Roque', 'Hockey', 'Jeu de paume', 'Rackets',
       'Skating', 'Water Motorsports', 'Modern Pentathlon', 'Ice Hockey',
       'Basketball', 'Canoe / Kayak', 'Handball', 'Judo', 'Volleyball',
       'Table Tennis', 'Badminton', 'Baseball', 'Softball', 'Taekwondo',
       'Triathlon', 'Canoe', 'Biathlon', 'Bobsleigh', 'Curling', 'Skiing',
       'Luge'], dtype=object)

In [87]:
#How many unique values are there?
oly["Sport"].nunique()

48

In [89]:
### When did boxing first appear as a sport in which women competed at the olympics?
oly.query ("Gender == 'Women' & Sport == 'Boxing'").sort_values(by = "Year", ascending = True)

Year    City   Sport Discipline        Athlete Name Athlete Country  \
29876  2012  London  Boxing     Boxing       ADAMS, Nicola             GBR   
29877  2012  London  Boxing     Boxing         REN, Cancan             CHN   
29878  2012  London  Boxing     Boxing     ESPARZA, Marlen             USA   
29879  2012  London  Boxing     Boxing           KOM, Mary             IND   
29896  2012  London  Boxing     Boxing       TAYLOR, Katie             IRL   
29897  2012  London  Boxing     Boxing     OCHIGAVA, Sofya             RUS   
29898  2012  London  Boxing     Boxing     ARAUJO, Adriana             BRA   
29899  2012  London  Boxing     Boxing   CHORIEVA, Mavzuna             TJK   
29912  2012  London  Boxing     Boxing   SHIELDS, Claressa             USA   
29913  2012  London  Boxing     Boxing  TORLOPOVA, Nadezda             RUS   
29914  2012  London  Boxing     Boxing           LI, Jinzi             CHN   
29915  2012  London  Boxing     Boxing     VOLNOVA, Marina             KAZ   

      Gender  Event   Medal  Season  
29876  Women  51 KG    Gold  Summer  
29877  Women  51 KG  Silver  Summer  
29878  Women  51 KG  Bronze  Summer  
29879  Women  51 KG  Bronze  Summer  
29896  Women  60 KG    Gold  Summer  
29897  Women  60 KG  Silver  Summer  
29898  Women  60 KG  Bronze  Summer  
29899  Women  60 KG  Bronze  Summer  
29912  Women  75 KG    Gold  Summer  
29913  Women  75 KG  Silver  Summer  
29914  Women  75 KG  Bronze  Summer  
29915  Women  75 KG  Bronze  Summer

## Summer v. Winter

It's likely that the Summer Olympics have more medal winners than Winter Olympics. How can we check?

In [91]:
## what is the total medal count for winter v. summer?
oly["Season"].value_counts()


Summer    31165
Winter     5770
Name: Season, dtype: int64

## How many medals were handed out at each olympics between 1896 and 2014?

Show the result from highest to lowest.

In [97]:
oly[["Year","Season"]].value_counts().sort_index(ascending = True)

Year  Season
1896  Summer     151
1900  Summer     512
1904  Summer     470
1908  Summer     804
1912  Summer     885
1920  Summer    1298
1924  Summer     884
      Winter     118
1928  Summer     710
      Winter      89
1932  Summer     615
      Winter     116
1936  Summer     875
      Winter     108
1948  Summer     814
      Winter     140
1952  Summer     889
      Winter     136
1956  Summer     885
      Winter     150
1960  Summer     882
      Winter     147
1964  Summer    1010
      Winter     185
1968  Summer    1031
      Winter     199
1972  Summer    1185
      Winter     200
1976  Summer    1305
      Winter     210
1980  Summer    1387
      Winter     218
1984  Summer    1459
      Winter     222
1988  Summer    1546
      Winter     264
1992  Summer    1705
      Winter     325
1994  Winter     343
1996  Summer    1859
1998  Winter     447
2000  Summer    2015
2002  Winter     481
2004  Summer    1998
2006  Winter     531
2008  Summer    2042
2010  Winter     529


## Filter and count

What are the top 5 sports that France has won the most gold in?

(HINT: different answers based on columns you decide to count)

In [104]:
gold_france = oly.query ("`Athlete Country` == 'FRA' & Medal == 'Gold'")
gold_france

Year       City     Sport        Discipline              Athlete Name  \
51     1896     Athens   Cycling     Cycling Track             FLAMENG, Léon   
54     1896     Athens   Cycling     Cycling Track              MASSON, Paul   
59     1896     Athens   Cycling     Cycling Track              MASSON, Paul   
62     1896     Athens   Cycling     Cycling Track              MASSON, Paul   
65     1896     Athens   Fencing           Fencing  GRAVELOTTE, Eugène-Henri   
...     ...        ...       ...               ...                       ...   
36270  2010  Vancouver    Skiing   Nordic Combined      LAMY CHAPPUIS, Jason   
36328  2014      Sochi  Biathlon          Biathlon          FOURCADE, Martin   
36332  2014      Sochi  Biathlon          Biathlon          FOURCADE, Martin   
36843  2014      Sochi    Skiing  Freestyle Skiing    CHAPUIS, Jean Frederic   
36917  2014      Sochi    Skiing         Snowboard          VAULTIER, Pierre   

      Athlete Country Gender                              Event Medal  Season  
51                FRA    Men                              100KM  Gold  Summer  
54                FRA    Men                               10KM  Gold  Summer  
59                FRA    Men                     1KM Time Trial  Gold  Summer  
62                FRA    Men                    Sprint Indivual  Gold  Summer  
65                FRA    Men                    Foil Individual  Gold  Summer  
...               ...    ...                                ...   ...     ...  
36270             FRA    Men  Individual, Ski Jumping K90 (70M)  Gold  Winter  
36328             FRA    Men                     12.5Km Pursuit  Gold  Winter  
36332             FRA    Men                               20KM  Gold  Winter  
36843             FRA    Men                          Ski Cross  Gold  Winter  
36917             FRA    Men                    Snowboard Cross  Gold  Winter  

[444 rows x 10 columns]

In [109]:
gold_france["Sport"].value_counts()

Fencing              116
Cycling               66
Handball              30
Aquatics              24
Equestrian            23
Skiing                22
Rowing                17
Football              17
Rugby                 17
Sailing               16
Athletics             14
Judo                  12
Shooting               9
Weightlifting          9
Biathlon               8
Tennis                 7
Canoe / Kayak          7
Archery                6
Skating                6
Boxing                 4
Wrestling              4
Croquet                4
Gymnastics             3
Canoe                  2
Water Motorsports      1
Name: Sport, dtype: int64